pkg/driver/node_server.go

```diff
@@ -486,80 +596,340 @@ func (driver *Driver) NodePublishVolume(ctx context.Context, request *csi.NodePu
+	// 		1) Create new volume using volume context parameters
+	// 		2) Controller publish volume (Add ACL)
+	// 		3) Stage volume on the node
+	//			-	Create device
+	//			- 	Mount to the target path using mount options and read-only flag)
+	//			-	Persist the staged device info within the POD directory path (targetPath's parent directory).
+	//		4) Publish volume on the node (Validates if device is mounted on the targetPath)
+
+	podUID := volumeContext[csiEphemeralPodUID]
+	// POD UID must be provided in the request
+	if podUID == "" {
+		log.Error("Missing POD uid in the volume context")
+		return status.Error(codes.InvalidArgument,
+			fmt.Sprintf("NodePublish of ephemeral volume %s failed due to missing POD uid in the request", volumeHandle))
+	}
+
+	var volCapability *csi.VolumeCapability
+
+	// Check if 'volumeModeBlock' is requested in the volume context.
+	if volumeContext[volumeModeBlockKey] == "true" {
+		log.Tracef("Ephemeral volume %s requested for block volume access", volumeHandle)
+
+		// Construct capability with AccessType as 'Block'
+		volCapability = &csi.VolumeCapability{
+			AccessType: &csi.VolumeCapability_Block{
+				Block: &csi.VolumeCapability_BlockVolume{},
+			},
+			AccessMode: volumeCapability.GetAccessMode(),
+		}
+	} else {
+		volCapability = volumeCapability
+	}
+
+	// Volume size (Default value will be used if 'sizeInGiB' parameter is unspecified)
+	size := int64(defaultVolumeSize)
+
+	// Get the volume size if requested in the volume context
+	sizeInGiB := volumeContext[sizeInGiBKey]
+	if sizeInGiB != "" {
+		log.Tracef("Ephemeral volume %s requested with size %s", sizeInGiB)
+
+		value, err := strconv.Atoi(sizeInGiB)
+		if err != nil {
+			return err
+		}
+		size = int64(value * GiBToByte) // Convert GiB to Bytes
+	}
+
+	// Append 'destroyOnDelete' to the volume context
+	volumeContext[destroyOnDeleteKey] = "true"
+
+	// Construct volume capabitilities to pass to createVolume()
+	volCapabilities := []*csi.VolumeCapability{
+		volCapability,
+	}
+
+	// Create volume
+	volume, err := driver.createVolume(
+		ephemeralVolName,
+		size,
+		volCapabilities,
+		secrets,
+		nil, /* No Volume Source */
+		volumeContext,
+		volumeContext, /* Pass volumeContext as-is for ephemeral volume */
+	)
+	if err != nil {
+		log.Error("err: ", err.Error())
+		return status.Error(codes.Internal,
+			fmt.Sprintf("Failed to create ephemeral volume %s, err: %s", ephemeralVolName, err.Error()))
+	}
+	log.Infof("Provisioned volume %s with ID %s for ephemeral volumeHandle %s", ephemeralVolName, volume.VolumeId, volumeHandle)
+
+	// Update DB entry
+	if err := driver.UpdateDB(dbKey, volume); err != nil {
+		return err
+	}
+
+	// Get Node Info
+	nodeID, err := driver.nodeGetInfo()
+	if err != nil {
+		log.Tracef("Failed to get node info, err: %s", err.Error())
+		return status.Error(codes.Internal,
+			fmt.Sprintf("Failed to node publish ephemeral volume %s due to error: %s", ephemeralVolName, err.Error()))
```

Comment:
don't we need a rollback if anything else fails after volume is created? I think nodeUnpublish will not be called to cleanup volume if nodePublish fails due to some error at some point.

Reply:
Agreed.  We should rollback if anything fails after creation.

Reply:
Should we look into using the command chains for create / execute /rollback ? https://github.com/hpe-storage/common-host-libs/blob/master/chain/chain.go

Reply:
> don't we need a rollback if anything else fails after volume is created? I think nodeUnpublish will not be called to cleanup volume if nodePublish fails due to some error at some point.

@shivamerla We don't need cleanup as it will be retried by the k8s(external-provisioner) whenever there is a failure of NodePublish(). Also, all CSI calls are implemented to be idempotent, so it will continued from where it is left of to complete the operation.

Reply:
@suneeth51 thats true, it keeps retrying to attach the volume, but it cannot go through for some reason(say multipath is not responding etc), they might end up deleting the pod. But since there is no PVC/PV associated with it, volume will be lying around. After this there is no way to delete the volume i think?  Let me know if i am missing something..

Reply:
Agreed. I will update the logic to handle the rollback/cleanup to remove the volume from the storage.